README.md
[中文](./README_zh.md)

# 📢Plugin Maintenance Notice

Due to the lack of updates and maintenance over an extended period of time, there may be various big and small issues with using this plugin.

## ⏫Upgrade OlivOS

To continue using the `BlackJack` plugin directly on `OlivOS`, we recommend upgrading `OlivOS` to the latest version and choosing to login with the Bilibili Danmaku protocol.

## 🐱‍🚀Join Group Chat

If you wish to continue using this plugin, you can join our group chat (Group ID: [`971050440`](https://jq.qq.com/?_wv=1027&k=VJqxAFTg)) and provide more detailed information. We will make efforts to fix it in our free time.

## BlackJack <img align="right" width="300" src="image/README/1682061505967.png">
* *Plugin based on [OlivOS](https://github.com/OlivOS-Team/OlivOS)*
* *Plugin based on [OlivaDiceCore](https://github.com/OlivOS-Team/OlivaDiceCore)*

> A simple interactive 21-point game based on OlivOS and OlivaDiceCore

![](https://img.shields.io/github/last-commit/HsiangNianian/BlackJack) [![CI](https://github.com/HsiangNianian/BlackJack/actions/workflows/ci.yml/badge.svg)](https://github.com/HsiangNianian/BlackJack/actions/workflows/ci.yml) [![](https://img.shields.io/github/downloads/HsiangNianian/BlackJack/total)](https://github.com/HsiangNianian/BlackJack/tags) [![](https://img.shields.io/github/v/release/HsiangNianian/BlackJack)](https://github.com/HsiangNianian/BlackJack/releases)

### 📕 21-point Gaming Rule Book

1. Game Objective

The aim of the game is to get your hand as close as possible to the value of 21 without going over OR to have a higher value hand than the dealer without exceeding 21.

2. Game Setup

- At the start of each game, the system generates a deck of cards and shuffles them.
- Players enter their bet and two cards are dealt to both the player and the dealer.
- The player can see their own cards and one of the dealer's card.

3. Game Process

- The player begins by deciding whether to hit (take another card) or stand (stay with current hand).
- If the player chooses to hit, the system will randomly draw another card from the deck and add it to the player's hand.
- If the value of the player's hand exceeds 21, the player busts and loses their bet.
- If the player chooses to stand, then it is the dealer's turn to decide whether to hit or stand.
- The dealer must take cards until the total is 17 or more points.
- If the dealer's hand exceeds 21, the dealer busts and the player wins the bet.
- If the dealer chooses to stand, then the hands of the dealer and the player are compared, and the higher hand wins. If the hands are equal, it is a tie.

4. Card Value Calculation

- Aces can count as either 1 or 11 points, face cards (J, Q, K) are worth 10 points, and all other cards are worth their face value.
- The sum of the value of cards in a player's hand is the player's score.
- The sum of the value of cards in the dealer's hand is the dealer's score.

5. Settlement Rules

- If the player's hand exceeds 21, the player busts and loses their bet.
- If the dealer's hand exceeds 21, the dealer busts and the player wins the bet.
- If the hands of the dealer and the player are both 21 or less, the hand closest to 21 wins.
- If the hands of the dealer and the player are equal, it is a tie.

6. End of Game

- After each game ends, the system will ask if you want to continue playing.
- If the player chooses to continue, the cards in the player's and dealer's hands will be cleared and a new game will begin.
- If the player loses all their bets, the game will end.

7. Command Parameters

`.BlackJack -n|--name [Player Name, Default: 无可救药的赌徒] -b|--bet [Initial Bet, Default: 1000]` //These can be omitted.

Enjoy the game!

### ⬇ Download

You can find the packaged and uncompressed `opk` plugin file for download in the [`release`](https://github.com/HsiangNianian/BlackJack/releases/latest) section on the right side.

Alternatively, you can also [`fork`](https://github.com/HsiangNianian/BlackJack/fork) this project and find the compressed file under `artifact` in the [`actions`](https://github.com/HsiangNianian/BlackJack/actions) section.

### 👨‍🚀 Contributors

<a href="https://github.com/HsiangNianian/BlackJack/graphs/contributors">
  <img width="50" src="https://contrib.rocks/image?repo=HsiangNianian/BlackJack" />
</a>

### 📄 License

[GPL 3.0](https://github.com/HsiangNianian/BlackJack/blob/main/LICENSE) © 2023-PRESENT [简律纯](https://github.com/HsiangNianian)


Changelog
### 0.1.0(4) 2023-04-21

- 修改中文简介

### 0.1.0(3) 2023-04-21

- 修复图片尺寸

### 0.1.0(2) 2023-04-21

- 多语种简介

### 0.1.0(1) 2023-04-21


BlackJack/main.py
# -*- encoding: utf-8 -*-
'''
     ██╗██╗   ██╗██╗   ██╗███╗   ██╗██╗  ██╗ ██████╗ 
     ██║╚██╗ ██╔╝██║   ██║████╗  ██║██║ ██╔╝██╔═══██╗
     ██║ ╚████╔╝ ██║   ██║██╔██╗ ██║█████╔╝ ██║   ██║
██   ██║  ╚██╔╝  ██║   ██║██║╚██╗██║██╔═██╗ ██║   ██║
╚█████╔╝   ██║   ╚██████╔╝██║ ╚████║██║  ██╗╚██████╔╝
 ╚════╝    ╚═╝    ╚═════╝ ╚═╝  ╚═══╝╚═╝  ╚═╝ ╚═════╝ 
                                                     
    Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
    https://www.apache.org/licenses/LICENSE-2.0
    Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
'''
import time
import random
import OlivOS
import BlackJack  # type: ignore
import OlivaDiceCore  #type: ignore
import argparse

parser = argparse.ArgumentParser(description='21点')
parser.add_argument('-n', '--name', default='无可救药的赌徒', help='玩家名称')
parser.add_argument('-b', '--bet', default=1000, help='初始赌注')

msg_id = None
bet = None
contextWaitTime = 60 * 2
contextFeq = 0.1
dictReplyContextReg = {}
# 卡牌和点数
suits = ['♠️', '♥️', '♣️', '♦️']
ranks = ['A', '2', '3', '4', '5', '6', '7', '8', '9', 'J', 'Q', 'K']
card_values = {'A': 11, '2': 2, '3': 3, '4': 4, '5': 5, '6': 6,
               '7': 7, '8': 8, '9': 9, 'J': 10, 'Q': 10, 'K': 10}
rule = """\
21点游戏规则书

1. 游戏目标

在不爆牌（手中的牌面点数超过21）的情况下，使自己手中的牌点数尽量接近21点，或者比庄家的牌点数更高。

2. 牌局准备

- 游戏开始时，系统会生成一副扑克牌，并将其打乱；
- 玩家需要输入赌注，然后系统会将两张牌发给玩家和庄家；
- 玩家可以看到自己手中的牌以及庄家的一张牌。

3. 游戏流程

- 玩家开始行动，可以选择“要牌”（继续抽牌）或“停牌”（不再抽牌）；
- 如果玩家选择“要牌”，系统会从扑克牌中随机抽出一张牌，并加入玩家手中的牌中；
- 如果玩家的牌面点数超过21点，则认为玩家已爆牌，游戏结束，玩家输掉赌注；
- 如果玩家选择“停牌”，则轮到庄家行动；
- 庄家会根据自己手中的牌面点数来决定是否继续抽牌；
- 如果庄家的牌面点数超过21点，则认为庄家已爆牌，玩家赢得赌注；
- 如果庄家选择“停牌”，则比较玩家和庄家的牌面点数，点数更高者胜利，如果点数相同则为平局。

4. 牌面点数计算

- A可以表示1或11点，2-10按照牌面点数计算，J、Q、K均算作10点；
- 玩家手中牌面点数之和即为玩家的点数；
- 庄家根据自己手中的牌面点数来决定是否继续抽牌，庄家手中牌面点数之和即为庄家点数。

5. 结算规则

- 如果玩家的牌面点数超过21点，则认为玩家已爆牌，游戏结束，玩家输掉赌注；
- 如果庄家的牌面点数超过21点，则认为庄家已爆牌，玩家赢得赌注；
- 如果玩家和庄家的牌面点数都不超过21点，则比较两者点数大小，点数更高者胜利；
- 如果玩家和庄家的牌面点数相同，则为平局。

6. 游戏结束

- 每次游戏结束后，系统会询问玩家是否继续游戏；
- 如果玩家选择继续，则清空玩家和庄家手中的牌，重新开始新一轮游戏；
- 如果玩家输掉了所有赌注，则游戏结束。

7. 指令参数

.BlackJack -n|--name [玩家名称，默认:无可救药的赌徒] -b|--bet [初始赌注，默认:1000] //这些都可以省略

愉快游戏！"""

class Event(object):
    def init(plugin_event, Proc):  # type: ignore
        global global_Proc
        global_Proc = Proc

    def group_message(plugin_event: OlivOS.API.Event, Proc: OlivOS.pluginAPI.shallow) -> None:  # type: ignore
        """
        Receive group messages
        """
        play_game(plugin_event, Proc)

    def private_message(plugin_event: OlivOS.API.Event, Proc: OlivOS.pluginAPI.shallow) -> None:  # type: ignore
        """
        Receive private messages
        """
        play_game(plugin_event, Proc)


def play_game(plugin_event: OlivOS.API.Event, Proc: OlivOS.pluginAPI.shallow):
    """开始游戏

    Args:
        plugin_event (OlivOS.API.Event): 插件事件
        Proc (OlivOS.pluginAPI.shallow): 进程接口
    """
    global bet

    class Deck:
        def __init__(self):
            self.cards = self._generate_deck()

        def _generate_deck(self):
            cards = []
            for suit in suits:
                for rank in ranks:
                    card = f'{rank}{suit}'
                    cards.append(card)
            return cards

        def shuffle(self):
            random.shuffle(self.cards)

        def draw_card(self):
            if not self.cards:
                # raise ValueError('牌堆里没有剩余的牌了。')
                eventReply(plugin_event, '牌堆里没有剩余的牌了。')
                return
            return self.cards.pop()

    class Hand:
        def __init__(self):
            self.cards = []

        def add_card(self, card):
            self.cards.append(card)

        def get_value(self):
            for card in self.cards:
                if card[0] not in card_values:
                    if 'qq' == plugin_event.platform['platform']:
                        try:
                            msg_id = plugin_event.data.message_id # type: ignore
                        except:
                            pass
                    eventReply(plugin_event, f'未知卡牌: {card}')
            value = sum([card_values[card[0]] for card in self.cards])
            num_aces = len([card for card in self.cards if card[0] == 'A'])
            while value > 21 and num_aces > 0:
                value -= 10
                num_aces -= 1
            return value

    class Player:
        def __init__(self, name, chips):
            self.name = name
            self.chips = int(chips)
            self.hand = Hand()

        def bets(self):
            global bet
            OlivaDiceCore.msgReply.replyMsg(
                plugin_event, f'{self.name}, 您有 {self.chips} 个筹码。您想下注多少？\n推荐的赌注: ({self.chips/2} {self.chips/5} {self.chips/10})')
            while True:
                input: 'str|None' = OlivaDiceCore.msgReplyModel.replyCONTEXT_regWait(
                    plugin_event=plugin_event,
                    flagBlock='allowCommand',
                    hash=OlivaDiceCore.msgReplyModel.contextRegHash(
                        [None, plugin_event.data.user_id]) # type: ignore
                )
                flag_need_loop = False
                if plugin_event != None:
                    flag_need_loop = True
                    if input is None:
                        eventReply(plugin_event, '无效的输入。请您输入一个数字。')
                    elif not isinstance(input, str):
                        eventReply(plugin_event, '无效的输入。请您输入一个数字。')
                    else:
                        try:
                            input_num = float(input)
                            if input_num <= self.chips:
                                if input_num <= 0:
                                    OlivaDiceCore.msgReply.replyMsg(
                                        plugin_event, '您不能下注小于0的点数。')
                                else:
                                    bet = input_num
                                    flag_need_loop = False
                                    break
                            else:
                                OlivaDiceCore.msgReply.replyMsg(
                                    plugin_event, '您不能下注超过您拥有的筹码数量。')
                        except ValueError:
                            eventReply(plugin_event, '无效的输入。请您输入一个数字。')
            self.chips -= bet
            return bet

        def hit_or_stand(self):
            OlivaDiceCore.msgReply.replyMsg(
                plugin_event, '您想要继续要牌(hit)还是停止要牌(stand)？ ')
            while True:
                action: 'str|None' = OlivaDiceCore.msgReplyModel.replyCONTEXT_regWait(
                    plugin_event=plugin_event,
                    flagBlock='allowCommand',
                    hash=OlivaDiceCore.msgReplyModel.contextRegHash(
                        [None, plugin_event.data.user_id]) # type: ignore
                ).lower()
                flag_need_loop = False
                if plugin_event != None:
                    flag_need_loop = True
                    if action == 'hit':
                        return 'hit'
                    elif action == 'stand':
                        return 'stand'
                    else:
                        eventReply(
                            plugin_event, '无效的输入。请您输入"hit"或者"stand"。')

    class Dealer:
        def __init__(self):
            self.hand = Hand()

        def hit_or_stand(self):
            if self.hand.get_value() < 17:
                return 'hit'
            else:
                return 'stand'

    message = plugin_event.data.message # type: ignore
    if message.split()[0].lower().startswith(".blackjack") or message.split()[0].lower().startswith("。blackjack"):
        if len(message.split()) >= 2:
            if message.split()[1].lower() == "rule":
                plugin_event.reply(rule)
                return
        args = parser.parse_args(message.split()[1:])
        deck = Deck()
        player = Player(args.name, args.bet)
        dealer = Dealer()
        while True:
            eventReply(plugin_event, '新一轮。')

            # 玩家下注
            bet = player.bets()

            # 发牌
            deck.shuffle()
            for i in range(2):
                player.hand.add_card(deck.draw_card())
                dealer.hand.add_card(deck.draw_card())

            # 玩家回合
            while True:
                plugin_event.reply(
                    f'{player.name}, 你的手牌: {", ".join(player.hand.cards)}\n点数: {player.hand.get_value()}')
                if player.hand.get_value() >= 21:
                    break
                action = player.hit_or_stand()
                if action == 'hit':
                    player.hand.add_card(deck.draw_card())
                else:
                    break

            # 庄家回合
            if player.hand.get_value() <= 21:
                while True:
                    time.sleep(random.random()*2)
                    plugin_event.reply(
                        f'庄家的手牌： {", ".join(dealer.hand.cards[:-1])}, *')
                    if dealer.hit_or_stand() == 'hit':
                        dealer.hand.add_card(deck.draw_card())
                    else:
                        break

            # 结算
            final = f'✨结算\n\n{player.name}的手牌: {", ".join(player.hand.cards)}\n玩家手牌点数总和: {player.hand.get_value()}\n\n庄家的手牌: {", ".join(dealer.hand.cards)}\n庄家手牌点数总和: {dealer.hand.get_value()}'

            if player.hand.get_value() > 21:
                plugin_event.reply(f'{final}\n\n你爆了。输了哦~')
            elif dealer.hand.get_value() > 21:
                plugin_event.reply(f'{final}\n\n庄家爆了，你赢了。')
                player.chips += bet * 2
            elif player.hand.get_value() > dealer.hand.get_value():
                plugin_event.reply(f'{final}\n\n你赢了！')
                player.chips += bet * 2
            elif player.hand.get_value() < dealer.hand.get_value():
                plugin_event.reply(f'{final}\n\n你输了。')
            else:
                plugin_event.reply(f'{final}\n\n平局！')
                player.chips += bet

            # 判断是否继续游戏
            plugin_event.reply(f'{player.name}, 您的筹码余额: {player.chips}\n你想要开始新的一局么？ (y/n) ')
            if player.chips <= 0:
                plugin_event.reply('您的筹码已经用完了。游戏结束。\n你想要开始新的一局么？ (y/n) ')
                break
            while True:
                flag_need_loop = False
                if plugin_event != None:
                    flag_need_loop = True
                    if flag_need_loop:
                        answer: 'str|None' = OlivaDiceCore.msgReplyModel.replyCONTEXT_regWait(
                                plugin_event=plugin_event,
                                flagBlock='allowCommand',
                            hash=OlivaDiceCore.msgReplyModel.contextRegHash(
                                [None, plugin_event.data.user_id]) # type: ignore
                        ).lower()
                        if answer == 'y':
                            player.hand = Hand()
                            dealer.hand = Hand()
                            break
                        elif answer == 'n':
                            plugin_event.reply('珍爱生命，远离赌博~3q!')
                            return
                        else:
                            eventReply(
                                plugin_event, '无效的输入，请输入 "y" 或者 "n"。')


def eventReply(plugin_event: OlivOS.API.Event, message: str, msgid: 'str|None' = None):
    if 'qq' == plugin_event.platform['platform']:
        try:
            msgid = plugin_event.data.message_id # type: ignore
        except:
            pass
    res = message
    if msg_id != None:
        res = '[OP:reply,id=%s]%s' % (str(msg_id), message)
    else:
        res = '[OP:reply,id=%s]%s' % (str(msgid), message)
    plugin_event.reply(res)
